repology/parser/slackware.py

```diff
@@ -0,0 +1,91 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with repology.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import sys
+
+from repology.package import Package
+
+
+class SlackwareParser():
+    def __init__(self):
+        pass
+
+# pylint: disable=R0201
+    def Parse(self, path):
+        packages = {}
+
+        # There is only The Man :)
+        volkerdi = ['volkerdi@slackware.com']
+
+        # Look for packages in these top level directories:
+        pkgdirs = ['slackware', 'slackware64', 'slackwarearm', 'patches', 'extra']
+        # Everything else is ignored.
+
+        with open(path, 'r') as filelisttxt:
+            for line in filelisttxt.readlines():
+                components = line.strip().split('/')
+                if len(components) < 4:
+                    # ignore lines that are not deep enough in the hierarchy (containing dirs etc)
+                    continue
+                topdir = components[1]
+                if topdir not in pkgdirs:
+                    continue
+                basename = components[-1]
+                if re.search('\.t.z$', basename):
+                    parsedbasename = basename.rsplit('-', 3)
+                    if len(parsedbasename) != 4:
+                        print('WARNING: unable to parse package name {}'.format(basename), file=sys.stderr)
+                        continue
+                    pkgnam = parsedbasename[0]
+                    pkgver = parsedbasename[1]
+
+                    # More than one instance of a package can be present.
+                    #   patches/  >  slackware*/  >  extra/
+
+                    if topdir == 'patches':
+                        pkgcat = 'patches'
+                        if pkgnam in packages:
+                            # overwrite the existing instance from slackware*/ or extra/
+                            packages[pkgnam].category = pkgcat
+                            packages[pkgnam].version = pkgver
+                    elif topdir == 'slackware' or topdir == 'slackware64' or topdir == 'slackwarearm':
+                        pkgcat = components[2]
+                        if pkgnam in packages:
+                            if packages[pkgnam].category == 'extra':
+                                # overwrite the existing instance from extra/ (e.g. nouveau blacklist)
+                                packages[pkgnam].category = pkgcat
+                                packages[pkgnam].version = pkgver
+                            else:
+                                # safe to assume the existing instance came from patches/, skip this instance in slackware*/
+                                pass
+                    elif topdir == 'extra':
+                        pkgcat = 'extra'
+                        if pkgnam in packages:
+                            # It must also be in patches/ or slackware*/, skip this instance in extra/
```

Comment:
There's at least one case for which this isn't true: `bash-completion` only exists in extra/